subxt/src/backend/unstable/follow_stream.rs

```diff
@@ -0,0 +1,311 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use super::rpc_methods::{FollowEvent, UnstableRpcMethods};
+use crate::config::Config;
+use crate::error::Error;
+use futures::{FutureExt, Stream, StreamExt};
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// This stream subscribes to `chainHead_follow` when polled, and
+/// resubscribes automatically if it's stopped.
+pub struct FollowStream<Hash> {
+    // Using this and not just keeping a copy of the RPC methods
+    // around means that we can test this in isolation with dummy streams.
+    stream_getter: FollowEventStreamGetter<Hash>,
+    stream: InnerStreamState<Hash>,
+}
+
+impl<Hash> std::fmt::Debug for FollowStream<Hash> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("FollowStream")
+            .field("stream_getter", &"..")
+            .field("stream", &self.stream)
+            .finish()
+    }
+}
+
+/// A getter function that returns an [`FollowEventStreamFut<Hash>`].
+pub type FollowEventStreamGetter<Hash> = Box<dyn FnMut() -> FollowEventStreamFut<Hash> + Send>;
+
+/// The future which will return a stream of follow events and the subscription ID for it.
+pub type FollowEventStreamFut<Hash> = Pin<
+    Box<dyn Future<Output = Result<(FollowEventStream<Hash>, String), Error>> + Send + 'static>,
+>;
+
+/// The stream of follow events.
+pub type FollowEventStream<Hash> =
+    Pin<Box<dyn Stream<Item = Result<FollowEvent<Hash>, Error>> + Send + 'static>>;
+
+/// Either a ready message with the current subscription ID, or
+/// an event from the stream itself.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum FollowStreamMsg<Hash> {
+    /// The stream is ready (and has a subscription ID)
+    Ready(String),
+    /// An event from the stream.
+    Event(FollowEvent<Hash>),
+}
+
+impl<Hash> FollowStreamMsg<Hash> {
+    /// Return an event, or none if the message is a "ready" one.
+    pub fn into_event(self) -> Option<FollowEvent<Hash>> {
+        match self {
+            FollowStreamMsg::Ready(_) => None,
+            FollowStreamMsg::Event(e) => Some(e),
+        }
+    }
+}
+
+enum InnerStreamState<Hash> {
+    /// We've just created the stream; we'll start Initializing it
+    New,
+    /// We're fetching the inner subscription. Move to Ready when we have one.
+    Initializing(FollowEventStreamFut<Hash>),
+    /// Report back the subscription ID here, and then start ReceivingEvents.
+    Ready(Option<(FollowEventStream<Hash>, String)>),
```

Comment:
Hehe, I sat pondering this for a bit myself!

The reason for the option is just because, below, we want to take ownership of those two values to put them in other places, but we only have a `&mut` reference to them. So the question is, if I want to take ownership of them, I have to replace the originals with something.. but what?! With an `Option` I can replace it with `None` using `option.take()` :)

I'll look at this again though because I can maybe do better here and it'd be lovely to remove the option!